examples/unfair_dice.py

```diff
@@ -0,0 +1,199 @@
+COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+"""
+
+# Part of this code comes from ptetools:
+# https://github.com/eendebakpt/ptetools/tree/main
+
+from functools import partial
+from typing import Any, Dict, List
+
+import numpy as np
+from opensquirrel.ir import Bit, Float, Qubit
+from qiskit_algorithms.optimizers import SPSA
+
+from quantuminspire.sdk.models.circuit import Circuit
+from quantuminspire.util.api.quantum_interface import ExecuteCircuitResult, QuantumInterface
+
+number_of_qubits = 2
+m = 2**number_of_qubits
+p0 = np.random.random(m) + 0.2
+p0 = p0 / np.sum(p0)
+target_distribution = {k: p0[k] for k in range(m)}
+
+
+def counts_to_distr(counts) -> dict[int, float]:
+    """Convert Qiskit result counts to a dictionary.
+
+    The dictionary has integers as keys, and pseudo-probabilities as values.
+    """
+    n_shots = sum(counts.values())
+    output_distr = {int(k, 2): v / n_shots for k, v in counts.items()}
+    return {k: output_distr[k] for k in sorted(output_distr)}
+
+
+class AverageDecreaseTermination:
+    def __init__(self, N: int, tolerance: float = 0.0):
+        """Callback to terminate optimization based the average decrease.
+
+        The average decrease over the last N data points is compared to the specified tolerance.
+        The average decrease is determined by a linear fit (least squares) to the data.
+
+        This class can be used as an argument to the Qiskit SPSA optimizer.
+
+        Args:
+            N: Number of data points to use
+            tolerance: Abort if the average decrease is smaller than the specified tolerance
+        """
+        self.N = N
+        self.tolerance = tolerance
+        self.reset()
+
+    @property
+    def parameters(self):
+        return self._parameters
+
+    @property
+    def values(self):
+        return self._values
+
+    def reset(self):
+        """Reset the data."""
+        self._values = []
+        self._parameters = []
+
+    def __call__(self, nfev, parameters, value, update, accepted) -> bool:
+        """
+        Args:
+            nfev: Number of evaluations
+            parameters: Current parameters in the optimization
+            value: Value of the objective function
+            update: Update step
+            accepted: Whether the update was accepted
+
+        Returns:
+            True if the optimization loop should be aborted
+        """
+        self._values.append(value)
+        self._parameters.append(parameters)
+
+        if len(self._values) > self.N:
+            last_values = self._values[-self.N :]
+            pp = np.polyfit(range(self.N), last_values, 1)
+            slope = pp[0] / self.N
+
+            if slope > self.tolerance:
+                return True
+        return False
+
+
+dt = AverageDecreaseTermination(N=35)
```

Comment:
This global variable seems a bit lost. I would prefer it if all global stuff is organised together. Maybe above the execute function?

Reply:
This is because it is also referred to in the finalize() function and there is currently no other way to share objects between the functions of the scripts running on the platform.

Reply:
Yes, I figured, but I would just like to have all the global variables in close proximity to each other in the code. And because dt can't move up (because forward declaration isn't a thing in Python) the other stuff (target_distribution and friends) will have to move down.

Reply:
Moved all globals toegther